File: Rendering/Matplotlib/Testing/Cxx/TestMathTextActor.cxx
/*=========================================================================

  Program:   Visualization Toolkit
  Module:    TestMathTextActor.cxx

  Copyright (c) Ken Martin, Will Schroeder, Bill Lorensen
  All rights reserved.
  See Copyright.txt or http://www.kitware.com/Copyright.htm for details.

     This software is distributed WITHOUT ANY WARRANTY; without even
     the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
     PURPOSE.  See the above copyright notice for more information.

=========================================================================*/

#include "vtkTextActor.h"

#include "vtkActor2D.h"
#include "vtkCamera.h"
#include "vtkCellArray.h"
#include "vtkCellData.h"
#include "vtkNew.h"
#include "vtkPolyData.h"
#include "vtkPolyDataMapper2D.h"
#include "vtkProperty2D.h"
#include "vtkRenderWindow.h"
#include "vtkRenderWindowInteractor.h"
#include "vtkRenderer.h"
#include "vtkTestingInteractor.h"
#include "vtkTextProperty.h"

#include <sstream>

namespace vtkTestMathTextActor {
void setupTextActor(vtkTextActor *actor, vtkPolyData *anchor)
{
  vtkTextProperty *p = actor->GetTextProperty();
  std::ostringstream label;
  label << p->GetVerticalJustificationAsString()[0]
        << p->GetJustificationAsString()[0] << " "
        << "$\\theta = " << p->GetOrientation() << "$";
  actor->SetInput(label.str().c_str());

  // Add the anchor point:
  double *pos = actor->GetPosition();
  double *col = p->GetColor();
  vtkIdType ptId = anchor->GetPoints()->InsertNextPoint(pos[0], pos[1], 0.);
  anchor->GetVerts()->InsertNextCell(1, &ptId);
  anchor->GetCellData()->GetScalars()->InsertNextTuple4(col[0] * 255,
                                                        col[1] * 255,
                                                        col[2] * 255, 255);
}
} // end namespace vtkTestMathTextActor3D

//----------------------------------------------------------------------------
int TestMathTextActor(int, char *[])
{
  using namespace vtkTestMathTextActor;
  vtkNew<vtkRenderer> ren;

  int width = 600;
  int height = 600;
  int x[3] = {100, 300, 500};
  int y[3] = {100, 300, 500};

  // Render the anchor points to check alignment:
  vtkNew<vtkPolyData> anchors;
  vtkNew<vtkPoints> points;
  anchors->SetPoints(points.GetPointer());
  vtkNew<vtkCellArray> verts;
  anchors->SetVerts(verts.GetPointer());
  vtkNew<vtkUnsignedCharArray> colors;
  colors->SetNumberOfComponents(4);
  anchors->GetCellData()->SetScalars(colors.GetPointer());

  for (size_t row = 0; row < 3; ++row)
  {
    for (size_t col = 0; col < 3; ++col)
    {
      vtkNew<vtkTextActor> actor;
      switch (row)
      {
        case 0:
          actor->GetTextProperty()->SetJustificationToRight();
          break;
        case 1:
          actor->GetTextProperty()->SetJustificationToCentered();
          break;
        case 2:
          actor->GetTextProperty()->SetJustificationToLeft();
          break;
      }
      switch (col)
      {
        case 0:
          actor->GetTextProperty()->SetVerticalJustificationToBottom();
          break;
        case 1:
          actor->GetTextProperty()->SetVerticalJustificationToCentered();
          break;
        case 2:
          actor->GetTextProperty()->SetVerticalJustificationToTop();
          break;
      }
      actor->GetTextProperty()->SetFontSize(22);
      actor->GetTextProperty()->SetOrientation(45.0 * (3 * row + col));
      actor->GetTextProperty()->SetColor(0.75, .2 + col * .26, .2 + row * .26);
      actor->GetTextProperty()->SetBackgroundColor(0.,
                                                   1. - col * .26,
                                                   1. - row * .26);
      actor->GetTextProperty()->SetBackgroundOpacity(0.25);
      actor->GetTextProperty()->SetFrame((row + col) % 9 == 0);
      actor->GetTextProperty()->SetFrameColor(
        col > 0 ? 1. : 0., col == 1 ? 1. : 0., col < 2 ? 1. : 0.);
      actor->GetTextProperty()->SetFrameWidth(1);
      actor->SetPosition(x[col], y[row]);
      setupTextActor(actor.GetPointer(), anchors.GetPointer());
      ren->AddActor(actor.GetPointer());
    }
  }

  vtkNew<vtkPolyDataMapper2D> anchorMapper;
  anchorMapper->SetInputData(anchors.GetPointer());
  vtkNew<vtkActor2D> anchorActor;
  anchorActor->SetMapper(anchorMapper.GetPointer());
  anchorActor->GetProperty()->SetPointSize(5);
  ren->AddActor(anchorActor.GetPointer());

  vtkNew<vtkRenderWindow> win;
  win->AddRenderer(ren.GetPointer());
  vtkNew<vtkRenderWindowInteractor> iren;
  iren->SetRenderWindow(win.GetPointer());

  ren->SetBackground(0.0, 0.0, 0.0);
  win->SetSize(width, height);

  win->SetMultiSamples(0);
  win->GetInteractor()->Initialize();
  win->GetInteractor()->Start();

  return EXIT_SUCCESS;
}

File: Rendering/Matplotlib/Testing/Python/TestMathTextActor.py
#!/usr/bin/env python

from vtkmodules.vtkCommonCore import (
    reference,
    vtkPoints,
    vtkUnsignedCharArray,
)

from vtkmodules.vtkCommonDataModel import (
    vtkCellArray,
    vtkCellData,
    vtkPolyData,
)

from vtkmodules.vtkRenderingCore import (
    vtkActor2D,
    vtkPolyDataMapper2D,
    vtkProperty2D,
    vtkRenderer,
    vtkRenderWindow,
    vtkRenderWindowInteractor,
    vtkTextActor,
    vtkTextProperty,
)

from vtkmodules.vtkRenderingFreeType import (
    vtkMathTextUtilities,
)

import vtkmodules.vtkRenderingMatplotlib
import vtkmodules.vtkRenderingOpenGL2
import vtkmodules.vtkRenderingUI

def setupTextActor(actor, anchor):
    p = actor.GetTextProperty()
    label = (str(p.GetVerticalJustificationAsString()[0]) +
             str(p.GetJustificationAsString()[0]) + " " +
             "$\\theta = " + str(round(p.GetOrientation())) + "$")
    actor.SetInput(label)

    # Add the anchor point:
    pos = actor.GetPosition()
    col = p.GetColor()
    ptId = reference((anchor.GetPoints().InsertNextPoint(pos[0], pos[1], 0.),))
    anchor.GetVerts().InsertNextCell(1, ptId)
    anchor.GetCellData().GetScalars().InsertNextTuple4(
        col[0] * 255, col[1] * 255, col[2] * 255, 255)

ren = vtkRenderer()

width = 600
height = 600
x = ( 100, 300, 500 )
y = ( 100, 300, 500 )

# Render the anchor points to check alignment:
anchors = vtkPolyData()
points = vtkPoints()
anchors.SetPoints(points)
verts = vtkCellArray()
anchors.SetVerts(verts)
colors = vtkUnsignedCharArray()
colors.SetNumberOfComponents(4)
anchors.GetCellData().SetScalars(colors)

for row in range(3):
    for col in range(3):
        actor = vtkTextActor()
        if row == 0:
            actor.GetTextProperty().SetJustificationToRight()
        elif row == 1:
            actor.GetTextProperty().SetJustificationToCentered()
        elif row == 2:
            actor.GetTextProperty().SetJustificationToLeft()
        if col == 0:
            actor.GetTextProperty().SetVerticalJustificationToBottom()
        elif col == 1:
            actor.GetTextProperty().SetVerticalJustificationToCentered()
        elif col == 2:
            actor.GetTextProperty().SetVerticalJustificationToTop()
        actor.GetTextProperty().SetFontSize(22)
        actor.GetTextProperty().SetOrientation(45.0 * (3 * row + col))
        actor.GetTextProperty().SetColor(0.75, .2 + col * .26, .2 + row * .26)
        actor.GetTextProperty().SetBackgroundColor(0., 1. - col * .26, 1. - row * .26)
        actor.GetTextProperty().SetBackgroundOpacity(0.25)
        actor.GetTextProperty().SetFrame((row + col) % 9 == 0)
        actor.GetTextProperty().SetFrameColor(
            (0., 1.)[col > 0], (0., 1.)[col == 1], (0., 1.)[col < 2])
        actor.GetTextProperty().SetFrameWidth(1)
        actor.SetPosition(x[col], y[row])
        setupTextActor(actor, anchors)
        ren.AddActor(actor)

anchorMapper = vtkPolyDataMapper2D()
anchorMapper.SetInputData(anchors)
anchorActor = vtkActor2D()
anchorActor.SetMapper(anchorMapper)
anchorActor.GetProperty().SetPointSize(5)
ren.AddActor(anchorActor)

renWin = vtkRenderWindow()
renWin.AddRenderer(ren)
iren = vtkRenderWindowInteractor()
iren.SetRenderWindow(renWin)

ren.SetBackground(0.0, 0.0, 0.0)
renWin.SetSize(width, height)

renWin.SetMultiSamples(0)
iren.Initialize()
if 'rtTester' not in locals() or rtTester.IsInteractiveModeSpecified():
    iren.Start()
